### PR TITLE
Implemented/integrated a basic state class

### PR DESCRIPTION
**Pull request by retr0h**
A class which manages the state file.  Intended to be used as a singleton
throughout molecule.  The classes' initial state is serialized to disk if the
file does not exist, otherwise is deserialized from the existing state file.
Changes made to the object are immediately serialized.

Fixes: #238

### CODE REVIEW

File: molecule/provisioners/vagrantprovisioner.py [hunk 1]
```diff
@@ -126,10 +128,10 @@ def default_platform(self):
                 0]['name']
 
         # default to first entry if no entry for platform exists or platform is false
-        if not self.m._state.get('default_platform'):
+        if not self.m._state.default_platform:
```
Comment (retr0h) [inline molecule/provisioners/vagrantprovisioner.py]: I'm not sure I understand why this is even necessary.  Why cant we simply look at `self.m_args['--platform']`

File: molecule/provisioners/vagrantprovisioner.py [hunk 2]
```diff
@@ -126,10 +128,10 @@ def default_platform(self):
...
             return default_platform
 
-        return self.m._state['default_platform']
+        return self.m._state.default_platform
```
Comment (retr0h) [inline molecule/provisioners/vagrantprovisioner.py]: Same here. :/

File: molecule/provisioners/vagrantprovisioner.py [hunk 3]
```diff
@@ -103,10 +105,10 @@ def default_provider(self):
                 0]['name']
 
         # default to first entry if no entry for provider exists or provider is false
-        if not self.m._state.get('default_provider'):
+        if not self.m._state.default_provider:
```
Comment (retr0h) [inline molecule/provisioners/vagrantprovisioner.py]: Why is this needed in a state file?

File: molecule/provisioners/dockerprovisioner.py [hunk 4]
```diff
@@ -161,7 +161,6 @@ def build_image(self):
                         tag_string))
```
Comment (retr0h) [inline molecule/provisioners/dockerprovisioner.py]: Why does the dockerprovisioner not use the state file, and use it's own status.

/cc @kireledan

**Comment by abrown-sg** [comment]
I believe the point of these was to maintain state after creating a machine. So that if you do `molecule converge --platform ubuntu` and then later do `molecule converge`, it will continue to use the platform that's currently running. Without it, the second `molecule converge` would stand up new instances with whatever the default platform is.  Trying to remember to add --platform to every converge call when you're developing a role is impossible.

**Comment by retr0h** [comment]
> I believe the point of these was to maintain state after creating an instance, and to persist it while the instances are running.

Ahhh, makes sense now.

**Comment by kireledan** [comment]
Definitely a much cleaner way of maintaining state. 👍

File: molecule/state.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+"""
+A class which manages the state file.  Intended to be used as a singleton
+throughout molecule.  The classes' initial state is serialized to disk if the
```
Comment (dhutty) [inline molecule/state.py]: I think you want "class's", since it's a singleton? Alternatively, rephrase for clarity?

File: molecule/commands.py [hunk 6]
```diff
@@ -186,12 +183,7 @@ def execute(self, exit=True):
         try:
             utilities.print_info("Destroying instances ...")
             self.molecule._provisioner.destroy()
-            self.molecule._state['default_platform'] = False
-            self.molecule._state['default_provider'] = False
-            self.molecule._state['created'] = False
-            self.molecule._state['converged'] = False
-            self.molecule._state['multiple_platforms'] = False
-            self.molecule._write_state_file()
+            self.molecule._state.reset()
```
Comment (abrown-sg) [inline molecule/commands.py]: ❤️ this!

File: molecule/state.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+#  THE SOFTWARE.
+"""
+A class which manages the state file.  Intended to be used as a singleton
+throughout molecule.  The initial state is serialized to disk if the file does
+not exist, otherwise is deserialized from the existing state file.  Changesmade
```
Comment (abrown-sg) [inline molecule/state.py]: Changesmade should be Changes made

**Comment by abrown-sg** [comment]
One small typo, otherwise this is marvelous! 👍
